src/components/ContactSelector/ContactSelectorOptions.js

```diff
@@ -0,0 +1,330 @@
+import { goToInvitationFlowAction } from 'actions/referralsActions';
+
+// Components
+import { Spacing } from 'components/Layout';
+import AddressScanner from 'components/QRCodeScanner/AddressScanner';
+import Button from 'components/Button';
+import ContactDetailsModal from 'components/ContactDetailsModal';
+import ContainerWithHeader from 'components/Layout/ContainerWithHeader';
+import EmptyStateParagraph from 'components/EmptyState/EmptyStateParagraph';
+import FloatingButtons from 'components/FloatingButtons';
+import ListItemWithImage from 'components/ListItem/ListItemWithImage';
+import Modal from 'components/Modal';
+import SearchBar from 'components/SearchBar';
+import SlideModal from 'components/Modals/SlideModal';
+
+// Selectors
+import { useRootSelector, activeAccountAddressSelector } from 'selectors';
+
+// Utils
+import { spacing } from 'utils/variables';
+import { getThemeColors } from 'utils/themes';
+import { getMatchingSortedData } from 'utils/textInput';
+import { resolveContact } from 'utils/contacts';
+import { isValidAddressOrEnsName } from 'utils/validators';
+import { addressesEqual } from 'utils/assets';
+
+// Types
+import type { Contact } from 'models/Contact';
+
+type Props = {|
+  contacts?: Contact[],
+  onSelectContact?: (contact: ?Contact) => mixed,
+  onResolvingContact?: (isResolving: boolean) => mixed,
+  allowCustomAddress?: boolean,
+  allowAddContact?: boolean,
+  title?: string,
+  searchPlaceholder?: string,
+|};
+
+const viewConfig = {
+  minimumViewTime: 300,
+  viewAreaCoveragePercentThreshold: 100,
+  waitForInteraction: true,
+};
+
+const MIN_QUERY_LENGTH = 2;
+
+const ContactSelectorOptions = ({
+  contacts = [],
+  onSelectContact,
+  onResolvingContact,
+  allowCustomAddress = true,
+  allowAddContact = true,
+  title = t('label.sendTo'),
+  searchPlaceholder = t('label.walletAddressEnsUser'),
+}: Props) => {
+  const theme = useTheme();
+
+  const searchInputRef = React.useRef<any>(null);
+  const modalRef = React.useRef<any>(null);
+
+  const [query, setQuery] = React.useState(null);
+  const [customAddressContact, setCustomAddressContact] = React.useState(null);
+  const [hasSearchError, setHasSearchError] = React.useState(false);
+
+  const dispatch = useDispatch();
+  const activeAccountAddress = useRootSelector(activeAccountAddressSelector);
+
+  const close = () => {
+    Keyboard.dismiss();
+    modalRef.current?.close();
+  };
+
+  const handleInputChange = (input: string) => {
+    input = input?.trim() ?? '';
+    setQuery(input);
+
+    if (allowCustomAddress) {
+      const isValid = isValidAddressOrEnsName(input);
+      setCustomAddressContact(isValid ? { name: input, ethAddress: input } : null);
+    }
+  };
+
+  const selectValue = async (contact: Contact) => {
+    close();
+
+    onResolvingContact?.(true);
+    const resolvedContact = await resolveContact(contact);
+    onResolvingContact?.(false);
+
+    onSelectContact?.(resolvedContact);
```

Comment:
Why pass `onResolvingContact` that changes state in the component above as a prop instead of calling `resolveContact` directly there in the function under `onSelectContact`?

Reply:
There reasons we're calling it is that first we close the `ContactSelectorOptions` modal returning to `ContactSelector` but in the meantime we start resolving ENS name, so `onResolvingContact` prop gives `ContactSelector` information when to display a spinner.

One alternative would be to call `close` after the resolution, but that gives bad UX when user selects an item and waits for ENS resolution and after couple of seconds the modal closes.

Reply:
I meant the alternative where `selectValue` calls `close` and `onSelectContact` with an _unresolved_ contact (and no waiting), and then resolution and spinner is handled in `<ContactSelector />` in order to keep the state inside one component instead of accessing it across two.

Reply:
Fair point, I've moved `resolveContact` call from `Options` to `ContactSelector`.